compiler/noirc_evaluator/src/ssa/parser.rs

```diff
@@ -0,0 +1,788 @@
+        let rhs = self.parse_value_or_error()?;
+        Ok(Some(ParsedInstruction::Constrain { lhs, rhs }))
+    }
+
+    fn parse_decrement_rc(&mut self) -> ParseResult<Option<ParsedInstruction>> {
+        if !self.eat_keyword(Keyword::DecRc)? {
+            return Ok(None);
+        }
+
+        let value = self.parse_value_or_error()?;
+        Ok(Some(ParsedInstruction::DecrementRc { value }))
+    }
+
+    fn parse_enable_side_effects(&mut self) -> ParseResult<Option<ParsedInstruction>> {
+        if !self.eat_keyword(Keyword::EnableSideEffects)? {
+            return Ok(None);
+        }
+
+        let condition = self.parse_value_or_error()?;
+        Ok(Some(ParsedInstruction::EnableSideEffectsIf { condition }))
+    }
+
+    fn parse_increment_rc(&mut self) -> ParseResult<Option<ParsedInstruction>> {
+        if !self.eat_keyword(Keyword::IncRc)? {
+            return Ok(None);
+        }
+
+        let value = self.parse_value_or_error()?;
+        Ok(Some(ParsedInstruction::IncrementRc { value }))
+    }
+
+    fn parse_range_check(&mut self) -> ParseResult<Option<ParsedInstruction>> {
+        if !self.eat_keyword(Keyword::RangeCheck)? {
+            return Ok(None);
+        }
+
+        let value = self.parse_value_or_error()?;
+        self.eat_or_error(Token::Keyword(Keyword::To))?;
+        let max_bit_size = self.eat_int_or_error()?.to_u128() as u32;
+        self.eat_or_error(Token::Keyword(Keyword::Bits))?;
+        Ok(Some(ParsedInstruction::RangeCheck { value, max_bit_size }))
+    }
+
+    fn parse_store(&mut self) -> ParseResult<Option<ParsedInstruction>> {
+        if !self.eat_keyword(Keyword::Store)? {
+            return Ok(None);
+        }
+
+        let value = self.parse_value_or_error()?;
+        self.eat_or_error(Token::Keyword(Keyword::At))?;
+        let address = self.parse_value_or_error()?;
+        Ok(Some(ParsedInstruction::Store { address, value }))
+    }
+
+    fn parse_assignment(&mut self, target: Identifier) -> ParseResult<ParsedInstruction> {
+        let mut targets = vec![target];
+
+        while self.eat(Token::Comma)? {
+            let target = self.eat_identifier_or_error()?;
+            targets.push(target);
+        }
+
+        self.eat_or_error(Token::Assign)?;
+
+        if self.eat_keyword(Keyword::Call)? {
+            let function = self.eat_identifier_or_error()?;
+            let arguments = self.parse_arguments()?;
+            self.eat_or_error(Token::Arrow)?;
+            let types = self.parse_types()?;
+            return Ok(ParsedInstruction::Call { targets, function, arguments, types });
+        }
+
+        if targets.len() > 1 {
+            return Err(ParserError::MultipleReturnValuesOnlyAllowedForCall {
+                second_target: targets[1].clone(),
+            });
+        }
+
+        let target = targets.remove(0);
+
+        if self.eat_keyword(Keyword::Allocate)? {
+            self.eat_or_error(Token::Arrow)?;
+            let typ = self.parse_type()?;
```

Comment:
Based on your illustration of how arrays are indexed now I remember that they are all flattened and index arithmetic is used to pick the right element, and when you call `allocate` you just get a pointer to the first element, which I assume could be a slice or an array. 

I notice that the parser accepts any type for an allocation, but the printer would [panic](https://github.com/noir-lang/noir/pull/6489/files#diff-58044f084c8337680eded1741fe40937b6a0e79d4250e5d1b9651de091de99c5R199) if it would be anything other than a reference type. If I'm not mistaken that `Type` is the same as this. Shouldn't this reject non-reference types to catch the problem before the print?

Reply:
Ah, no problem there, I see that `FunctionBuilder::insert_allocate` wraps this type into a reference 👍

Reply:
So you're suggesting that instead of this:
```
v1 = allocate -> [Field; 5]
```
it would look like this:
```
v1 = allocate -> &mut [Field; 5]
```

Speaking as a newcomer to the codebase, I didn't know all references were mutable; for example looking at this SSA, I thought e.g. `&[Field; 3]` would be an option:
```
brillig(inline) fn push f1 {
  b0(v0: &mut [Field; 3], v1: &mut u32, v2: Field):
```
I finally found out from the print format itself that the `"&mut"` part comes from the [type display format](https://github.com/noir-lang/noir/blob/b65a63d8d898e46cc686baa500f0b8070e45df14/compiler/noirc_evaluator/src/ssa/ir/types.rs#L229), there is no other option. Obviously in the context of a function parameter where it could be passed by value or by mutable reference this differentiation is crucial. 

In the case of `allocate` it's an interesting conundrum. If we all know that `allocate` allocates an `&mut` and the input is a `Type`, then I think we shouldn't have to type the `&mut` just for the sake of it, but the `->` syntax suggests that `allocate` "returns" this thing, which again if you know what `allocate` does, then it makes sense to you, otherwise it might not. For example `allocate::<[Field; 5]>` makes me recognise that it's an input, not an output.

So I agree that `allocate -> &mut <type>` would be more intuitive, because that's how the `Type::Reference` would print itself anyway, and that's what `allocate` inserted, so that's what it should return. Whether people would be annoyed by having to type extra characters I don't know.

Reply:
I'll make this change and also improve error messages before merging.